spl/labels.py

```diff
@@ -186,55 +198,90 @@ def __get_application_numbers(self, set_id, bs_content):
-                                ) in SplHistoricalLabels.LABEL_SECTIONS[
-                                    section_title
-                                ]:
-                                    if sub_section_title in sub_title.text:
-                                        # Extract the sub-title text
-                                        sub_sections.append(
-                                            {
-                                                "name": sub_section_title,
-                                                "text": get_xml_text(
-                                                    sub_title.parent
-                                                ),
-                                            }
-                                        )
-                                        # Remove the sub title from the main title
-                                        sub_title.parent.decompose()
-
-                        # Add the section text, along with the sub-sections
+        i = 0
+        while i < len(titles):
+            title = titles[i]
+            # test if any substring of title is in LABEL_SECTION
+            title_substring_in_LABEL_SECTION = any(
+                label_section.lower() in get_xml_title(title.text).lower()
+                for label_section in SplHistoricalLabels.LABEL_SECTIONS
+            )
+
+            # test if any string matches is in LABEL_SECTIONS_EXACT
+            title_string_in_LABEL_SECTION_EXACT = any(
+                label_section.lower() == get_xml_title(title.text).lower()
+                for label_section in SplHistoricalLabels.LABEL_SECTIONS_EXACT
+            )
+
+            # test if any substring is in LABEL_SECTION_BLACKLIST
+            title_substring_in_LABEL_SECTION_BLACKLIST = any(
+                label_section.lower() in get_xml_title(title.text).lower()
+                for label_section in SplHistoricalLabels.LABEL_SECTIONS_BLACKLIST
+            )
+
+            if (
+                title_substring_in_LABEL_SECTION
+                or title_string_in_LABEL_SECTION_EXACT
+            ) and not title_substring_in_LABEL_SECTION_BLACKLIST:
+                subtitles = title.parent.find_all("title")
+                if len(subtitles) > 1:
+                    # if title has subtitles, add title with sibling <text>
+                    parent = get_xml_title(title.text)
+                    sibling_text = (
+                        title.find_next_sibling("text").text
+                        if title.find_next_sibling("text")
+                        else ""
+                    )
+                    labels.append(
+                        {
+                            "name": parent,
+                            "text": sibling_text,
+                            "parent": None,
+                        }
+                    )
+                    # loop through all subtitles and add subtitle and text
+                    for j in range(1, len(subtitles)):
```

Comment:
For subtitle in subtitles ... ?

Reply:
I would keep this as is.  Subtitles technically includes the parent title, because of `subtitles = title.parent.find_all("title")`.  So the first subtitle is the parent title.  Hence looping from the second subtitle to the last one.  I believe this is a quirk with how the xml is arranged.  The best way to grab subtitles is to go to parent and get are titles within parent, and exclude the first one because it is the title of the parent.  An alternative to grab titles is to do child sibling child recursions, which is more complex.

Reply:
Just made a simple change here: `for subtitle in subtitles[1:]`